From c4a207faedde63ba25ea46f3c06b1821712e9d9b Mon Sep 17 00:00:00 2001
From: Pavel Grafov <pgrafov@google.com>
Date: Mon, 7 Jul 2025 12:58:01 +0100
Subject: [PATCH] Prevent settings impostors from installing CA certs

In addition to checking that the invoking package was
com.android.settings, check that this package has system uid.

Bug: 417463103
Test: manual
Flag: EXEMPT bugfix
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:46580f2f6f027dfe2284382184785f2979677303)
Merged-In: I3779b0d38dafee2333716113f6d5f4de78529380
Change-Id: I3779b0d38dafee2333716113f6d5f4de78529380
---
 .../android/certinstaller/CertInstaller.java  | 25 +++++++++++++++++--
 .../certinstaller/CredentialHelper.java       |  5 ----
 2 files changed, 23 insertions(+), 7 deletions(-)

diff --git a/src/com/android/certinstaller/CertInstaller.java b/src/com/android/certinstaller/CertInstaller.java
index 570c400..883c897 100644
--- a/src/com/android/certinstaller/CertInstaller.java
+++ b/src/com/android/certinstaller/CertInstaller.java
@@ -30,6 +30,7 @@ import android.content.pm.PackageManager;
 import android.os.AsyncTask;
 import android.os.Bundle;
 import android.os.Process;
+import android.os.UserHandle;
 import android.security.Credentials;
 import android.security.KeyChain;
 import android.security.KeyChain.KeyChainConnection;
@@ -69,6 +70,7 @@ public class CertInstaller extends Activity {
 
     // key to states Bundle
     private static final String NEXT_ACTION_KEY = "na";
+    private static final String SETTINGS_PACKAGE = "com.android.settings";
 
     private final ViewHelper mView = new ViewHelper();
 
@@ -241,7 +243,7 @@ public class CertInstaller extends Activity {
                 new Pkcs12ExtractAction("").run(this);
             }
         } else {
-            if (mCredentials.calledBySettings()) {
+            if (isCalledBySettings()) {
                 MyAction action = new InstallOthersAction();
                 action.run(this);
             } else {
@@ -250,6 +252,25 @@ public class CertInstaller extends Activity {
         }
     }
 
+    private boolean isCalledBySettings() {
+        if (!SETTINGS_PACKAGE.equals(mCredentials.getReferrer())) {
+            return false;
+        }
+
+        // On some systems there is no preinstalled "com.android.settings". To block settings
+        // impostors from installing CA certificates, we check referrer's uid.
+        final PackageManager pm = getPackageManager();
+        final int settingsUid;
+        try {
+            settingsUid = pm.getApplicationInfo(SETTINGS_PACKAGE, 0 /* flags */).uid;
+        } catch (PackageManager.NameNotFoundException e) {
+            Log.e(TAG, "Unable to find info for settings package");
+            return false;
+        }
+
+        return UserHandle.isSameApp(settingsUid, Process.SYSTEM_UID);
+    }
+
     private class InstallVpnAndAppsTrustAnchorsTask extends AsyncTask<Void, Void, Boolean> {
 
         @Override
@@ -353,7 +374,7 @@ public class CertInstaller extends Activity {
         removeDialog(PROGRESS_BAR_DIALOG);
         if (success) {
             removeDialog(PKCS12_PASSWORD_DIALOG);
-            if (mCredentials.calledBySettings()) {
+            if (isCalledBySettings()) {
                 if (validCertificateSelected()) {
                     installCertificateOrShowNameDialog();
                 } else {
diff --git a/src/com/android/certinstaller/CredentialHelper.java b/src/com/android/certinstaller/CredentialHelper.java
index eab58f5..4274a22 100644
--- a/src/com/android/certinstaller/CredentialHelper.java
+++ b/src/com/android/certinstaller/CredentialHelper.java
@@ -71,7 +71,6 @@ class CredentialHelper {
     private static final String DATA_KEY = "data";
     private static final String CERTS_KEY = "crts";
     private static final String USER_KEY_ALGORITHM = "user_key_algorithm";
-    private static final String SETTINGS_PACKAGE = "com.android.settings";
 
     private static final String TAG = "CredentialHelper";
 
@@ -296,10 +295,6 @@ class CredentialHelper {
         return mCertUsageSelected;
     }
 
-    boolean calledBySettings() {
-        return mReferrer != null && mReferrer.equals(SETTINGS_PACKAGE);
-    }
-
     Intent createSystemInstallIntent(final Context context) {
         Intent intent = new Intent("com.android.credentials.INSTALL");
         // To prevent the private key from being sniffed, we explicitly spell
-- 
2.52.0.rc1.455.g30608eb744-goog

